syft/pkg/cataloger/nix/parse_nix_store_path_test.go

```diff
@@ -0,0 +1,284 @@
+package nix
+
+import (
+	"path"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_findVersionIsh(t *testing.T) {
+	tests := []struct {
+		name    string
+		input   string
+		wantIdx int
+		wantEx  string
+		wantPr  string
+	}{
+		{
+			name:    "no version",
+			input:   "5q7vxm9lc4b9hifc3br4sr8dy7f2h0qa-source",
+			wantIdx: -1,
+			wantEx:  "",
+			wantPr:  "",
+		},
+		{
+			name:    "semver with overbite into output",
+			input:   "/nix/store/h0cnbmfcn93xm5dg2x27ixhag1cwndga-glibc-2.34-210-bin",
+			wantIdx: 50,
+			wantEx:  "2.34-210-bin",
+			wantPr:  "210-bin",
```

Comment:
`bin` is just the name of the output, it's not part of the version. There can be multiple outputs with the same "version number", but different output names. I'd probably just drop the optional -{bin,lib,dev,doc,devdoc} (these are the most common ones) output name from the version number - or is there a notion of multiple package outputs in syft?

Reply:
Question: is `glibc-2.34` really the "package" and the single package has multiple outputs? If that's the case, there is more work to be done -- the cataloger should be merging these into a single package, and the metadata for each needs to account for expressing multiple outputs (and surrounding metadata).

Reply:
This is the `bin` output for the glibc 2.34 build. It includes binaries produced by the build, like `bin/getent`. The glibc .so file would be in a separate output (`lib` or the default output, where the prefix is omitted).

It makes sense to distinguish between these outputs (so the info should not be merged), and it also makes sense to keep the output hash as an identifier.

We can have multiple glibc-2.34 in the same "closure" / filesystem (let's say one carrying a patch, the other not), and we can ship the `bin` outputs or not, depending on whether we want `getent` or not.

If there's a vulnerability, the output_path allows distinguishing between the patched version or unpatched version, and knowing which outputs you ship also helps determining whether you might be affected by a certain vulnerability and it's limited to the CLI parser, not the library.

I'm having a hard time calling this package, because it's nothing bundled up, and these are the output paths produced by a build. After évaluation, Nix uses "derivations" as build recipes. Each derivation is a single build, and it may produce multiple outputs (with output paths).

Reply:
It sounds like the best way forward is to collect these definitions into a single package, where the metadata would be able to track the multiple outputs:
```json
{
  "name": "glibc"
  "version": "2.34"
  "metadata": {
    "outputs": [
        {
          "output": "man"
          "outputhash": "cfmmhxh0c5nxnag1bixdg7d3g2c9wan2"
          "files": [...]
        }
        {
          "output": "bin"
          "outputhash": "h0cnbmfcn93xm5dg2x27ixhag1cwndga"
          "files": [...]
        }
      ]
  }
}
```

In this way there is still a single logical `glibc@2.34` package and grype downstream would be able to look at the `outputhash` for each output on the metadata.

> If there's a vulnerability, the output_path allows distinguishing between the patched version or unpatched version

it sounds like the `output_path` would be really important to try and capture on the metadata directly then? Just to double check... is this output path the path to the store (e.g. `/nix/store/hs0yi5n5nw6micqhy8l1igkbhqdkzqa1-foo`)? Or is this referring to another value?

> I'm having a hard time calling this package, because it's nothing bundled up, and these are the output paths produced by a build.

I can see what you're saying here (I think). I started with the assumption of what you might find on search.nixos.org ... say for glibc https://search.nixos.org/packages?channel=22.11&show=glibc&from=0&size=1&sort=relevance&type=packages&query=glibc 
<img width="937" alt="Screenshot 2023-03-27 at 11 07 07 AM" src="https://user-images.githubusercontent.com/590471/227981925-237dd60e-4393-4967-ab75-2be43c578dda.png">

I feel this orients the results in terms of the logical name + version and outputs available. I've been equating the name + version as the idea of a singular logical package and the list of outputs installed as something that is attached metadata to that (single) package. I see what you're saying with nothing is being bundled up but syft doesn't necessarily require that packaging ecosystems do any bundling of builds/files at all.

Is this a kosher way to think about "packages" in the Nix ecosystem? Or said another way, could representing singular packages with possibly multiple outputs be inaccurate or misleading in downstream analysis?

Reply:
> it sounds like the output_path would be really important to try and capture on the metadata directly then? Just to double check... is this output path the path to the store (e.g. /nix/store/hs0yi5n5nw6micqhy8l1igkbhqdkzqa1-foo)? Or is this referring to another value?

Let me pick a better example:

If I take a current checkout of nixpkgs (4bb072f0a8b267613c127684e099a70e1f6ff106), there's multiple outputs of the `openssl` package, described in the "build recipe" (.drv):

```
$ nix repl .
nix-repl> openssl
«derivation /nix/store/3jbv3scpmj2kxhshfz544jq37dzywbss-openssl-3.0.8.drv»

nix-repl> openssl.outputs
[ "bin" "dev" "out" "man" "doc" "debug" ]

nix-repl> :b openssl

This derivation produced the following outputs:
  bin -> /nix/store/05qna7girdhnzx38jkhr4yyh7gmnizv3-openssl-3.0.8-bin
  debug -> /nix/store/spa8fwjwd6r4n6d4gqg4s1ylijmkc0x7-openssl-3.0.8-debug
  dev -> /nix/store/ggsmy3rwsr6q6ri45ck3jdqnz9cf6dm6-openssl-3.0.8-dev
  doc -> /nix/store/bsabnghysgdmfhjlj2pal3i25q9w2brf-openssl-3.0.8-doc
  man -> /nix/store/4ac38nfgh4whllggv4zwp14va4mfirr7-openssl-3.0.8-man
  out -> /nix/store/s8vg2h8xzqmjd72f3g3p1jqy2lbbapc6-openssl-3.0.8
```

So, all these output paths are produced by the same build recipe (`/nix/store/3jbv3scpmj2kxhshfz544jq37dzywbss-openssl-3.0.8.drv`).

 - The `bin` output of that derivation has an output hash of `05qna7girdhnzx38jkhr4yyh7gmnizv3` ([nixbase32](https://cs.tvl.fyi/depot/-/blob/tvix/nix-compat/src/nixbase32.rs?L1) encoding).
 - The `debug` output has an output hash of `spa8fwjwd6r4n6d4gqg4s1ylijmkc0x7`
 - The `dev` output has an output hash of `ggsmy3rwsr6q6ri45ck3jdqnz9cf6dm6`
 - ...

If I now go back in time to a previous nixpkgs checkout (2 weeks before, 1df7332c722cc8d235b979cfa8f1bbe949b722fb), I still have the same openssl version in that case, but the build recipe is a different one (and due to this, all the hashes).

```
$ nix repl .
nix-repl> openssl   
«derivation /nix/store/y43ka1zh6icrf515hc5qzb643wsscxmd-openssl-3.0.8.drv»

nix-repl> openssl.outputs
[ "bin" "dev" "out" "man" "doc" "debug" ]

nix-repl> :b openssl

This derivation produced the following outputs:
  bin -> /nix/store/b1rd646j6yc22qnhiydb5a1hi5hc4dky-openssl-3.0.8-bin
  debug -> /nix/store/8zw06pv5v5l7ggklyls77n5fw03vgdr7-openssl-3.0.8-debug
  dev -> /nix/store/gs1841l5x53q3r5l196ld9kxq45pvf82-openssl-3.0.8-dev
  doc -> /nix/store/42krirjplr0a4z3w2myhrawg5kbgmy40-openssl-3.0.8-doc
  man -> /nix/store/rrz6hdzxsxg3yadk02gcypib36bq2zqm-openssl-3.0.8-man
  out -> /nix/store/fgrj06y1x83fwh8hqgg02v9abc7a7b65-openssl-3.0.8
```

Now the `bin`output hash is `b1rd646j6yc22qnhiydb5a1hi5hc4dky`.

Both these packages can coexist in /nix/store.

In more complicated closures / "container contents", it's very well possible for, let's say, two slightly different flavors of libraries to be linked against two different binaries.

Key takeaways:
 - just the "name" (openssl-3.0.8) doesn't uniquely identify a package, the outputhash is crucial
 - The output hash uniquely identifies an output, as soon as the build recipe (or any of its dependencies) changes it'll differ
 - Just by looking at the paths, without any additional database (that's what syft is doing) it's not possible to map / group certain output paths to the .drv / build recipe producing it.

Reply:
Thanks for the example / input here -- much appreciated! I think the takeaways make sense given the fact that syft is guessing from a directory structure instead of package metadata information. In the future I feel it would be ideal to improve this in some way to make it more obvious which outputs are related... but I don't see a safe way to do this given the information available and the non-guarentees on the sqlite cache that's there. Thanks for helping to navigate these traps @flokli !

Ok, back to your question (this was all a little bit of a tangent):

> bin is just the name of the output, it's not part of the version. There can be multiple outputs with the same "version number", but different output names. I'd probably just drop the optional -{bin,lib,dev,doc,devdoc} (these are the most common ones) output name from the version number - or is there a notion of multiple package outputs in syft?

This function first extracts the version as if it were a semver, then later in processing (in `parseNixStorePath`) the last non-numeric, dash-delimited string is assumed to be the nix output. That is, the package ultimately reflects your suggestion (to not include output names in the version).

There isn't a notion of a "package output" in syft, however, what it will look like to most users is that there are multiple instances of the package `name-version` installed, and only when they look at the `.metadata` or `.purl` will they see the specific output this package represents.